src/controllers/deploy.go

```diff
@@ -69,15 +69,26 @@ func internaldeploy(a *history.ActionInstance) ([]byte, error) {
 		state.Access = a.Access
 		state.Server = a.Server
 		state.Status = "deploying"
-		history.SetState(a.RepoURL, state)
-		output, err = exec.Command(deployScriptName, "-n", "-u", a.RepoURL, "-b", branch, "-m", a.Server, "-s", a.Subdomain, "-a", a.Access).CombinedOutput()
+		mux.Lock()
+		if history.CheckSubdomain(state.Subdomain) {
+			history.SetState(a.RepoURL, state)
```

Comment:
I would suggest to check the domain in the history.SetState() function. This way it will be atomic. Let that function return an error if the state cannot be set. 

Then you don't need  a mux in deploy.go

Reply:
Something like:
```
if err1 != history.SetState(a.RepoURL,state); err1 != nil {
    set ouptut and err
    return output, err
}
output, err = exec.Command(.....)
.....
```